joomla/joomla-cms
# [6.2] Add a global master language for Multilingual Associations

This adds an optional global master language for the multilingual associations. The topic that had its beginnings in the two PR's #17881 and #17802 was tackled from scratch.

This gives each association a "Master-Child" relationship. This means that an associated item set with the master language is set as master and the items in the other languages are set as children.
This makes it possible to add new association statuses to the already existing 'not associated' and 'associated' ones: the 'outdated' and 'up-to-date' status. These new statuses maintain the child elements in relation to their master. 
When associations are created, the children are given up-to-date status. If the master item is changed, the children will be set to outdated (if they were up-to-date until then). A distinction is made between whether the content history is enabled or not. The difference: content history only saves entries in which changes are really made. 
In the Associations List View a new filter for the Association States has been added.

To update an outdated child to up-to-date a new edit view is called, in which it is only possible to save and update the target (i.e. the outdated child). 
Here, too, a distinction is made between whether the content history is activated or not. If content history is activated, a customized Compare View of contenthistory is displayed. The changes of the master are displayed since the moment the child was set to outdated and to the current master.
If it is content history disabled, the normal edit view of the reference/master is displayed. The user has to determine whether there are changes to the master or not. Advantage over before: by the status outdated you at least know that someone has saved the master.

### Summary of Changes
- The Associations List View is the only List View in the backend that has the behavior that no content is displayed the first time this view is called. This is confusing, especially for users who have not yet spent long time with associations. To adapt this list view to the others so that a more uniform behavior is achieved in the backend, the filters for itemtype and language are filled with default values. This means that the two buttons 'delete all associations' and 'delete orphans' must always be displayed accordingly (otherwise they were hidden as soon as a list appears - see #21869). Delete All Associations no longer deletes all associations, but all associations with matching context.
- The colors for the association states 'associated' and 'not associated' have now changed everywhere: gray for 'not associated' and green for associated and 'up-to-date'.  Orange now stands for 'outdated'.
- The option to set the global master language has been added to the languagefilter plugin.
- New columns in #__associations: 
    - master_id' in which the ID of the master item is stored for the child items otherwise 0 for the master item itself and 
    - 'master_date' where either the modified date or the save date of the history table of the master item is stored. The child item gets the same master date as its master item if they are newly associated or when the child item gets updated. 
- Fix bug where newsfeeds and contacts get the same assoc key. 
To prevent associations from receiving the same association key, since id and language can be the same in different contexts, the context is also added as differentiation. 
As the bug appears only after my new queries (key query only), this is fixed here.
- For the masterlanguage see above.


### Testing Instructions
- Install a new instance because there are changes in the sql.
- Install at least two languages. For example, install the multilingual sample data and then the normal sample data (so there is sample data for associated and non-associated items).
- Now no global master language is set yet. The behaviour of the associations should be the same as before. Only the colors have changed. 
No empty list appears in the Associations List View and the delete buttons are now always displayed.
- Go to the Languagefilter plugin and activate a global master language.
Go back to the Associations List View. A new filter appears and the associations are displayed differently, also in other List Views with associations. 
- Change a master item. With versions enabled (default) you need to really make changes and save the master in order to check if children gets 'outdated'. 
Update a child element. 


### Expected result
- Associations List View without a global Master Language:
![image](https://user-images.githubusercontent.com/31204883/60507515-9cd35a80-9cc8-11e9-86f4-b2d55433723e.png)
- Articles List View without a global Master Language:
![image](https://user-images.githubusercontent.com/31204883/60507613-d906bb00-9cc8-11e9-8632-8af9180d449c.png)

- Set a global Master Language in the Languagefilter Plugin:
![SetGlobalMaster](https://user-images.githubusercontent.com/31204883/60506941-474a7e00-9cc7-11e9-9e07-e8bd7758ca06.gif)
- Associations List View with Global Master Language:
![image](https://user-images.githubusercontent.com/31204883/60507255-056e0780-9cc8-11e9-9cf4-730836e4d2b9.png)
- Articles List View with Global Master Language:
![image](https://user-images.githubusercontent.com/31204883/60507319-2cc4d480-9cc8-11e9-9b25-3ba5d8549e4d.png)
- New Update/Edit View for outdated items with versions enabled:
![image](https://user-images.githubusercontent.com/31204883/60507807-487caa80-9cc9-11e9-998a-db5e3d86da0e.png)


### Actual result
- No global Master Language Setting in the Languagefilter Plugin.
- No 'outdated' or 'up-to-date' association states.
- Associations List View:
![FirstEntryAssociations](https://user-images.githubusercontent.com/31204883/60507394-58e05580-9cc8-11e9-9b2f-2a53be154cf3.gif)
- Articles List View:
![image](https://user-images.githubusercontent.com/31204883/60507569-bffe0a00-9cc8-11e9-9cac-ceb4ee3e46d4.png)


### Documentation Changes Required
Yes
